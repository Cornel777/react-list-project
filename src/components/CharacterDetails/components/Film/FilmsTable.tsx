import { TableContainer, Paper, Table, TableHead, TableBody } from "@mui/material";
import { StyledTableRow, StyledTableCell } from "../styled/styledTable";
import Film from "./Film";
import { FunctionComponent } from "react";

interface FilmsTableProps {
  data: {
    films: string[];
  }
}

const FilmsTable: FunctionComponent<FilmsTableProps> = ({ data }) => {
  return (
    <TableContainer component={Paper} style={{ maxWidth: '400px' }}>
    <Table size="medium" aria-label="Films table">
     <TableHead>
        <StyledTableRow>
          <StyledTableCell>Film name</StyledTableCell>
           <StyledTableCell>Episode no.</StyledTableCell>
        </StyledTableRow>
      </TableHead>
      <TableBody>
        {data.films.map((filmUrl: string) => {
          return <Film filmUrl={filmUrl} />
        })}
      </TableBody>
    </Table>
  </TableContainer>
  )
}

export default FilmsTable;