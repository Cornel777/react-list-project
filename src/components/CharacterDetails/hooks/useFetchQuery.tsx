import { useQuery, keepPreviousData } from "@tanstack/react-query";
import fetchSwapi from "../../../services/api/fetchSwapi";

export interface HookReturn<T> {
  isFetching: boolean;
  isPlaceholderData: boolean;
  isPending: boolean;
  isError: boolean;
  error: Error | null;
  data: T | undefined;
}


const useFetchQuery = <T, >(queryKey: string, path: string, id: string | undefined): HookReturn<T> => {

  const { isFetching, isPlaceholderData, isPending, isError, error, data } = useQuery<T>({
    queryKey: [queryKey, id],
    queryFn: () => fetchSwapi(path),
    staleTime: 5000,
    placeholderData: keepPreviousData,
  })

    return {
      isFetching,
      isPlaceholderData,
      isPending,
      isError,
      error,
      data,
    }
}

export default useFetchQuery;