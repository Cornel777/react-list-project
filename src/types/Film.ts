import { URL } from './Character';

export interface Film {
  title: string;
  episode_id: number;
  opening_crawl: string;
  director: string;
  producer: string;
  release_date: string;
  characters: URL[];
  planets: URL[];
  starships: URL[];
  vehicles: URL[];
  species: URL[];
  created: Date;
  edited: Date;
  url: URL;
}