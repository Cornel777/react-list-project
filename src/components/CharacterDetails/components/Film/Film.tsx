import { FunctionComponent } from "react"
import LoadingDots from "../../../LoadingDots/LoadingDots"
import { type Film } from "../../../../types/Film"
import { StyledTableCell, StyledTableRow } from "../styled/styledTable"
import useFetchQuery from "../../hooks/useFetchQuery"

interface FilmProps {
  filmUrl: string
}

const Film: FunctionComponent<FilmProps> = ({filmUrl}) => {

  const filmUrlList = filmUrl.split('/').filter((element: string) => element.length > 0);
  const filmId = filmUrlList.pop();

  const { isPending, isError, error, data} = useFetchQuery<Film>('film', `films/${filmId}/`, filmId)

  if (isPending) {
    return <LoadingDots />
  }
  if (isError) {
    return <p>{`Failed to load films ${error && error.message}`}</p>
  }
  if (!data) {
    return null;
  }

  return (
    <StyledTableRow key={filmId}>
      <StyledTableCell> {data.title}</StyledTableCell>
      <StyledTableCell>Episode {data.episode_id}</StyledTableCell>
    </StyledTableRow>
  )
}

export default Film;