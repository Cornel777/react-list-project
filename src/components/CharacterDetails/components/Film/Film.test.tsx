import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import Film from './Film';
import useFetchQuery from '../../hooks/useFetchQuery';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

vi.mock('../../hooks/useFetchQuery');
// vi.mock('../../../LoadingDots/LoadingDots', () => ({
//   default: () => <div data-testid="loading-dots">Loading...</div>
// }));

const mockFilmData = {
  title: 'A New Hope',
  episode_id: 4,
  director: 'George Lucas',
  producer: 'Gary Kurtz, Rick McCallum',
  release_date: '1977-05-25',
  opening_crawl: 'It is a period of civil war...',
  url: 'https://swapi.dev/api/films/1/',
  characters: [],
  planets: [],
  starships: [],
  vehicles: [],
  species: [],
  created: '',
  edited: ''
};

const mockFilmUrl = 'https://swapi.dev/api/films/1/';

describe('Film component', () => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false
      }
    }
  });

  beforeEach(() => {
    vi.clearAllMocks();
    vi.mocked(useFetchQuery).mockReturnValue({
      isPending: false,
      isError: false,
      isFetching: false,
      isPlaceholderData: false,
      error: null,
      data: mockFilmData
    });
  });



  function renderWithClient(ui: React.ReactElement) {
    const testQueryClient = queryClient;
    const { rerender, ...result } = render(
        <QueryClientProvider client={testQueryClient}>{ui}</QueryClientProvider>
    )
    return {
        ...result,
        rerender: (rerenderUi: React.ReactElement) =>
            rerender(
                <QueryClientProvider client={testQueryClient}>{rerenderUi}</QueryClientProvider>
            ),
    }
  }

  it('matches the snapshot', () => {
    const { asFragment } = renderWithClient(<Film filmUrl={mockFilmUrl} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders film details', () => {
    renderWithClient(<Film filmUrl={mockFilmUrl} />);
    expect(screen.getByText('A New Hope')).toBeInTheDocument();
    expect(screen.getByText('Episode 4')).toBeInTheDocument();
    expect(useFetchQuery).toHaveBeenCalledWith('film', 'films/1/', '1');
  });

  it('renders loading dots', () => {
    vi.mocked(useFetchQuery).mockReturnValue({
      isPending: true,
      isError: false,
      isFetching: false,
      isPlaceholderData: false,
      error: null,
      data: undefined
    });

    renderWithClient(<Film filmUrl={mockFilmUrl} />);
    expect(screen.getByTestId('loading-dots')).toBeInTheDocument();
  });

  it('returns null if no data is available', () => {
    vi.mocked(useFetchQuery).mockReturnValue({
      isPending: false,
      isError: false,
      isFetching: false,
      isPlaceholderData: false,
      error: null,
      data: undefined
    });

    renderWithClient(<Film filmUrl={mockFilmUrl} />);
    expect(screen.queryByText('A New Hope')).not.toBeInTheDocument();
    expect(screen.queryByText('Episode 4')).not.toBeInTheDocument();
  });

  it('renders error message', async () => {
    vi.mocked(useFetchQuery).mockReturnValue({
      isPending: false,
      isError: true,
      isFetching: false,
      isPlaceholderData: false,
      error: new Error('Error loading film'),
      data: undefined
    });

    renderWithClient(<Film filmUrl={mockFilmUrl} />);
    expect(await screen.findByText(/Failed to load films/i)).toBeInTheDocument();
  });
});