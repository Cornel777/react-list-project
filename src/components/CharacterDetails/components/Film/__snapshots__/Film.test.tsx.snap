// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Film component > matches the snapshot 1`] = `
<DocumentFragment>
   A New HopeEpisode 4
</DocumentFragment>
`;
